### ISSUE TITLE
Feature: implied parentheses

### ISSUE THREAD

**Issue by deepfire** [lexi-lambda/hackett#22]
Since this is 21'th century, we probably can have many (most?) parentheses implied, in a way that is not dissimilar to how Haskell does implicit blocks.

To consider the difference of presentation -- using the example from the README:
```
#lang hackett

data Maybe a
  nothing
  just a

def x : Int
  let y 3
       z 7
    y + z

class Show a
  show : Int -> a -> String

instance forall [a] (Show a) => (Show (Maybe a))
  def+ show
    nothing  -> "nothing"
    just x -> "just " <> show x
```
..versus:
```
#lang hackett

(data (Maybe a)
  nothing
  (just a))

(def x : Int
  (let ([y 3]
        [z 7])
    {y + z}))

(class (Show a)
  [show : {Int -> a -> String}])

(instance (forall [a] (Show a) => (Show (Maybe a)))
  (def+ show
    [nothing  -> "nothing"]
    [(just x) -> {"just " <> (show x)}]))
```

**Comment by deepfire** [comment]
As a personal perspective -- after long time of doing Lisp, and then coming to Haskell, I have come to value its visual terseness.

It seems to increase the raw perception bandwidth.

**Comment by lexi-lambda** [comment]
I agree that this would be nice if it could be done perfectly, but it’s trickier in practice than it might seem. I *do* think it would be cool to have an opt-in ML-like reader that you could get with `#lang hackett/ml` or something like that. For prior art, check out the following SRFIs:

- [SRFI 49: Indentation-sensitive syntax](https://srfi.schemers.org/srfi-49/srfi-49.html)
- [SRFI 110: Sweet-expressions](https://srfi.schemers.org/srfi-110/srfi-110.html) — This one has an [existing Racket implementation](http://docs.racket-lang.org/sweet/index.html) that works with arbitrary `#lang`s that use the Racket reader, *including* (as of now) Hackett.
- [SRFI 119: wisp](https://srfi.schemers.org/srfi-119/srfi-119.html)

The fact that there are three different standards for this hints that this is a harder problem than it might seem. I would encourage you to study the different tradeoffs each of them makes to better understand what the challenges are.

**Comment by johanatan** [comment]
While I agree that Haskell-like syntax can be easier on the eyes, I doubt that the individual perception bandwidth is increased all that much. And the parentheses are not only useful for macros but for editors' structural editing capabilities (which can be thought of as development-time [i.e., pre-compile time] key-bound macros).
